prover/src/constraints/evaluator/default.rs

```diff
@@ -107,8 +110,28 @@ where
         #[cfg(debug_assertions)]
         evaluation_table.validate_transition_degrees();
 
-        // combine all evaluations into a single column and return
-        evaluation_table.combine()
+        // combine all evaluations into a single column
+        let combined_evaluations = {
+            let main_and_aux_evaluations = evaluation_table.combine();
+
+            match self.evaluate_lagrange_kernel_constraints(trace, domain) {
+                Some(lagrange_kernel_combined_evals) => {
+                    // if present, linearly combine the Lagrange kernel evaluations too
+                    debug_assert_eq!(
+                        main_and_aux_evaluations.len(),
+                        lagrange_kernel_combined_evals.len()
+                    );
+                    main_and_aux_evaluations
+                        .into_iter()
+                        .zip(lagrange_kernel_combined_evals)
+                        .map(|(eval_1, eval_2)| eval_1 + eval_2)
+                        .collect()
+                }
+                None => main_and_aux_evaluations,
+            }
```

Comment:
This is probably fine for now, but we may be able to optimize this a bit more in the future. Specifically: right now we first compute the entire vector with Lagrange constraint evaluations, and then add the values to the main/aux evaluations. By refactoring this a bit (and passing `main_and_aux_evaluations` into the Lagrange constraint evaluator we could avoid the second loop (i.e., we'd treat `main_and_aux_evaluations` as an accumulator to which Lagrange constraint evaluations get added immediately as they are computed for each row).

Let's add a `TODO` here and also create an issue for this.

Reply:
I implemented it and got the exact same performance (within noise bounds):

```
trace column building: 84ms
precompute trans divisors time: 11ms
invert trans divisors time: 34ms
invert boundary divisors time: 21ms
evaluate and merge constraints time: 404ms

TOTAL lagrange constraints eval time: 472ms
```

Since the first one is more readable and in a functional style, I would suggest we revert that [commit](https://github.com/facebook/winterfell/pull/247/commits/7c74d894a835f3734da38cb2cee02723605cb7d9). What do you think?

Reply:
I don't have a strong preference here - so, fine with either of the approaches.